web: Add `ruffle()` (Version 1) API

Dinnerbone opened the pull request:
Bikeshedding time, here's the chance to decide what the "v1 api" looks like before we lock it in.

The only change so far is that I've added a specific method for calling EI, vs the very easy-to-break and implicit "the method will exist, trust me". I'm awaiting feedback before touching anything else.

Whilst this is "version 1" (`PlayerV1` interface), if we never add a V2 that's totally fine. `ruffle()` is the same as `ruffle(1)`.

Things to consider for each existing member of the API:
- Is anything a property that should be a method?
- Is anything a method that should be a property?
- Should we rename anything, or change arguments?

Dinnerbone commented on web/packages/core/src/internal/player/impl_v1.ts:
```diff
@@ -0,0 +1,109 @@
+        this.#inner = inner;
+    }
+
+    get onFSCommand(): ((command: string, args: string) => boolean) | null {
```
3 possibilities here:
1. As today, we set **the** callback for **all** FS commands.
2. We change it to register **a** callback for **all** FS commands.
3. We change it to register **the** callback for **a specific named** FS command.

I think I kinda prefer 2.

Dinnerbone commented on web/packages/core/src/internal/player/impl_v1.ts:
```diff
@@ -0,0 +1,109 @@
+        this.#inner.onFSCommand = value;
+    }
+
+    get readyState(): ReadyState {
```
We should probably use an actually public enum, maybe string based

Dinnerbone commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+     * @param args A string passed to the host application for any use.
+     * @returns True if the command was handled.
+     */
+    onFSCommand: ((command: string, args: string) => boolean) | null;
```
3 possibilities here:

As today, we set the callback for all FS commands.
We change it to register a callback for all FS commands.
We change it to register the callback for a specific named FS command.

I think I kinda prefer 2.

evilpie replied:
We could also consider switching the player to be a proper `EventTarget` and use `addEventListener` for this.

Dinnerbone replied:
If we go `EventTarget` we'll lose the warning of unhandled commands, unless we require people to flag that they've handled it. Right now we ask that the `onFSCommand` returns true when handled.

I think that's probably fine, but what do you think?

Dinnerbone replied:
I thought about this a bit and decided that event may not be the way to go. A few issues:
1. Versioning will be difficult
2. It's a bit weird to have the element have the event when everything else is behind `ruffle()` - people may accidentally `ruffle().addEventListener`
3. Typescript really does not seem to like custom events... makes it super verbose to use :(

I'll go with option 2 for now and we can reevaluate in the future for a V2.

Dinnerbone commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+     *
+     * @returns The `ReadyState` of the player.
+     */
+    get readyState(): ReadyState;
```
We should probably use an actually public enum, maybe string based

danielhjacobs replied:
Probably, keep in mind the idea here may generally be to match https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/readyState#value (though HAVE_CURRENT_DATA, HAVE_FUTURE_DATA, and HAVE_ENOUGH_DATA don't necessarily all make sense for Ruffle, though depending how streamed loading of content is implemented maybe they will).

Dinnerbone replied:
Is this to match it for compatibility, or just for being a similar thing? Because if it's the latter; there's also document.readyState which _is_ a string, for precedence

danielhjacobs replied:
Context is in https://github.com/ruffle-rs/ruffle/pull/4084 and https://github.com/ruffle-rs/ruffle/pull/4813#issuecomment-892177759

Dinnerbone commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+    /**
+     * Plays or resumes the movie.
+     */
+    play(): void;
```
This is actually Ruffle's suspend/resume, not Flash's play/pause. I suggest we rename this to `resume()` or `suspended = false` or something similar

Dinnerbone commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+     */
+    play(): void;
+
+    /**
```
This is actually Ruffle's suspend/resume, not Flash's play/pause. I suggest we rename this to `isSuspended`

Dinnerbone commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+     *
+     * @returns True if this player is playing, false if it's paused or hasn't started yet.
+     */
+    get isPlaying(): boolean;
```
This is actually Ruffle's suspend/resume, not Flash's play/pause. I suggest we rename this to isSuspended

Dinnerbone commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+     *
+     * @param isFull Whether to set to fullscreen or return to normal.
+     */
+    setFullscreen(isFull: boolean): void;
```
Maybe a setter?

evilpie replied:
Why do we have this method and enter/exitFullscreen?

evilpie replied:
Oh sorry, I just saw your comment below.

Maybe we should have a getter and setter `fullscreen`, and a getter for `fullscreenAllowed`?

Dinnerbone replied:
I think I'd prefer actions/verbs to be methods, not just a property (it feels weird, not sure I can articulate it, it's like a side effect even if it's documented as such?)

How about we just go with `requestFullscreen()`/`exitFullscreen` to match the same methods on `Element`?

Dinnerbone commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+     *
+     * This is not guaranteed to succeed, please check [[fullscreenEnabled]] first.
+     */
+    enterFullscreen(): void;
+
+    /**
+     * Requests the browser to no longer make this player fullscreen.
+     */
+    exitFullscreen(): void;
```
These are identical to `setFullscreen(true)` / `setFullscreen(false)`, seems a bit redundant

Dinnerbone commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+     * No more frames, scripts or sounds will be executed.
+     * This movie will be considered inactive and will not wake up until resumed.
+     */
+    pause(): void;
```
This is actually Ruffle's suspend/resume, not Flash's play/pause. I suggest we rename this to `suspend` or `suspended = true` or something

Dinnerbone commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+     *
+     * @param observer The observer that will be called for each trace.
+     */
+    set traceObserver(observer: ((message: string) => void) | null);
```
Do we want to set **the** trace observer, as today - or set **a** trace observer?

Dinnerbone replied:
I think setting is fine.

evilpie commented on web/packages/core/src/public/v1.ts:
```diff
@@ -0,0 +1,173 @@
+     *
+     * @returns True if you may call [[enterFullscreen]].
+     */
+    get fullscreenEnabled(): boolean;
```
Isn't this name misleading? I think this could easily be confused to with "isFullscreen".

danielhjacobs replied:
Doesn't it match https://developer.mozilla.org/en-US/docs/Web/API/Document/fullscreenEnabled?

Dinnerbone replied:
tbh I don't entirely agree with having this method. It's a document thing, not at all related to this player or element. The source is `return document.fullscreenEnabled || document.webkitFullscreenEnabled` and whilst I guess webkit prefix is certainly a consideration... is it our place? It just feels weird to have a method like that.

Dinnerbone commented:
I've gone ahead and changed `pause()`/`play()`/`isPlaying` to `suspend()`/`resume()`/`suspended` as that seems fairly agreed upon. Comment if anyone disagrees with the specifics before this PR is finished.

Dinnerbone commented:
Renamed `enterFullscreen` to `requestFullscreen`, and removed `setFullscreen`.
Decided to keep `fullscreenEnabled` as if we still support old ios, it's likely users will forget to check the prefixed stuff too.

Dinnerbone commented:
Changed `onFSCommand` (getter/setter) to `addFSCommandHandler` (method).

Dinnerbone commented:
Made `ReadyState` a string

danielhjacobs commented on web/packages/core/src/public/player/legacy.ts:
```diff
@@ -1,56 +1,106 @@
+ *
+ * @deprecated Use {@link PlayerV1.readyState} to get a {@link ReadyState} instead.
+ */
+export enum LegacyReadyState {
```
Can we remove this enum and just compare any legacy ready states to numbers instead of enum members? This only adds confusion.

Dinnerbone replied:
I'm not sure that having people just need to know what arbitrary numbers means is better than code constants (which an enum basically is). That seems even more confusing to me.
Adrian argued in favour of just keeping the legacy enum and not having the new one, which whilst I'm not personally in favour of, I think I'd prefer over switching and entirely not documenting the old style :(

Dinnerbone replied:
Reverted to old style API, and made the old enum public
